### PR TITLE
PXB-2106 : copy-back creates wrong binlog.index

### PR DESCRIPTION
**Pull request by kennt-percona**
PXB-2457 : Incorrect binlog names if binlog name contains periods

Issue
When performing copy-back, PXB will add incorrect file names to the binlog index.
This occurs when:
(1) an absolute path is used for the binlog and
(2) when the path is the datadir path.

Also, if the binlog name contains a period ('.'), we were not building
the binlog file names the same way as MySQL.

Solution
This occurs because the system treats the path as an absolute path
(because it is specified as an absolute path) and it treats it as
a relative path (because the path is in the datadir).
This exposed a bug in how we were parsing the paths.

Binlog names with a period ('.') are not generated the same way as MySQL.
Everything after the first '.' is removed.

### CODE REVIEW

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 1]
```diff
@@ -1589,15 +1589,24 @@ struct binlog_file_location {
 
       if (opt_log_bin != nullptr) {
         if (Fil_path::is_absolute_path(opt_log_bin)) {
-          r.name =
+          r.name = std::string(opt_log_bin).substr(dirname_length(opt_log_bin));
+
+          /* Truncate at the first '.' like MySQL */
+          r.name = r.name.substr(0, r.name.find(FN_EXTCHAR)) + suffix;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: This same code is used in the else branch. Can we move `r.name` to before  `if (Fil_path::is_absolute_path(opt_log_bin)) {` and inside the if/else compute only `r.path`

Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: Move the declaration of `suffix` to this part of the code too.

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 2]
```diff
@@ -1589,15 +1589,24 @@ struct binlog_file_location {
...
           fn_format(buf, opt_log_bin, datadir.c_str(), "", MY_UNPACK_FILENAME);
-          r.name = opt_log_bin + suffix;
-          r.path = buf + suffix;
+
+          r.name = std::string(opt_log_bin).substr(dirname_length(opt_log_bin));
+
+          /* Truncate at the first '.' like MySQL */
+          r.name = r.name.substr(0, r.name.find(FN_EXTCHAR)) + suffix;
+
+          r.path = std::string(buf).substr(0, dirname_length(buf)) + r.name;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc:1619]: Can you please let me know why this part is needed?
If understand correctly, the issue only happens if you have a full path that points to datadir. This else branch falls into the not absolute path.

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
A few inline comments plus:

* Is this 8.0 only? If not, please start with 2.4 path
* Can you please adjust the commit message to include the link to jira ticket, like https://github.com/percona/percona-xtrabackup/commit/80c1ba62aa65d9b0583ca2ad0eb21616b45264a2
* Can you please run jenkins against it. For PXB we use https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/ to trigger a compile + test tasks. After ready, please send us the link for your test param (https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/)

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 2 above)
Comment (kennt-percona) [reply storage/innobase/xtrabackup/src/backup_copy.cc:1619]: This is the other bug that was being fixed.  PXB-2457.

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 1 above)
Comment (kennt-percona) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: code moved

**Comment by kennt-percona** [comment]
@altmannmarcelo 
This is 8.0 only.

**Review by altmannmarcelo — APPROVED** [review]
@kennt-percona Thanks for fixing my comments. LGTM

**Comment by Techshrimp** [comment]
This is due to the fact that the binlog name is different before and after the backup and causes problems during parsing, which can be replicated normally, and is related to the target_location function.

from percona-xtrabackup-8.0.6 to percona-xtrabackup-8.0.23-16

binlog_file_location target_location(const std::string &datadir) const {
    binlog_file_location r;

    if (!name.empty()) {
      std::string suffix = fn_ext(name.c_str());

      if (opt_log_bin != nullptr) {
        if (Fil_path::is_absolute_path(opt_log_bin)) {
          r.name =
              std::string(opt_log_bin).substr(0, dirname_length(opt_log_bin)) +
              suffix;
          r.path = opt_log_bin + suffix;
        } else {
          char buf[FN_REFLEN];
          fn_format(buf, opt_log_bin, datadir.c_str(), "", MY_UNPACK_FILENAME);
          r.name = opt_log_bin + suffix;
          r.path = buf + suffix;
        }

**Pull request merged**